File: 2010_2011_mf_sujet_6_ks_regression_lineaire/regr2.m
function [ bhat,sigma ] = regr2(Y,MX )
%UNTITLED11 Summary of this function goes here
%   Detailed explanation goes here
n=length(Y)
k=size(MX,2)
u=ones(1,n)';
M2=[u,MX];
g=(M2'*M2)^(-1);
bhat=g*M2'*Y;
sigma=norm(Y-M2*bhat)/sqrt(n-k)


end

File: 2010_2011_mf_sujet_6_ks_regression_lineaire/regr2.py
import numpy as np


def regr2(Y, MX):
    """Multiple linear regression of Y on columns of MX (intercept added automatically)."""
    n = len(Y)
    k = MX.shape[1]
    M2 = np.column_stack([np.ones(n), MX])
    g = np.linalg.inv(M2.T @ M2)
    bhat = g @ M2.T @ Y
    sigma = np.linalg.norm(Y - M2 @ bhat) / np.sqrt(n - k)
    print(f'bhat = {bhat}')
    print(f'sigma = {sigma}')
    return bhat, sigma
